File: day2.py
#!/usr/bin/env python3

def is_report_safe(report):
    if report[0] == report[1]:
        return False
    report_is_increasing = report[0] < report[1]
    for level, next_level in zip(report, report[1:] + [None]):
        if next_level is None:
            return True
        level_difference = next_level - level
        if report_is_increasing:
            if level_difference < 1 or level_difference > 3:
                return False
        if not report_is_increasing:
            if level_difference < -3 or level_difference > -1:
                return False

def part1(reports):
    nb_safe = 0
    for report in reports:
        if is_report_safe(report):
            nb_safe += 1
    print(nb_safe)

def part2(reports):
    nb_safe = 0
    for report in reports:
        if is_report_safe(report):
            nb_safe += 1
            continue
        for i in range(len(report)):
            dampened_report = list(report)
            del (dampened_report[i])
            if is_report_safe(dampened_report):
                nb_safe += 1
                break

    print(nb_safe)

reports = []

with open("day2.txt", "r") as input:
    for line in input:
        reports.append(line.split())
        reports[-1] = [int(n) for n in reports[-1]]

part1(reports)
part2(reports)

File: day2/day2.cpp
// The challenge for this one was not to use bruteforce

#include <iostream>
#include <fstream>
#include <ostream>
#include <sstream>
#include <vector>
#include <string>
#include <tuple>

std::tuple<bool, unsigned int> is_report_safe(const std::vector<unsigned short>& report){
    std::vector<int> differences;
    int sumDiff = 0;
    for (unsigned int i = 0; i < report.size() - 1; i++){
        differences.push_back(static_cast<int>(report[i + 1]) - static_cast<int>(report[i]));
        sumDiff += (differences.back() >= 0) ? 1 : -1;
    }
    bool isIncreasing = (sumDiff >= 0) ? true : false;
    for (unsigned int i = 0; i < report.size() - 1; i++){
        if ((isIncreasing and (differences[i] < 1 or differences[i] > 3))
        or (!isIncreasing and (differences[i] < -3 or differences[i] > -1))){
            return {false, i};
        }
    }
    return {true, 0};
}

int part1(const std::vector<std::vector<unsigned short>>& reports){
    unsigned int accumulator = 0;
    for (const std::vector<unsigned short>& report : reports){
        if (std::get<bool>(is_report_safe(report))){
            accumulator++;
        }
    }
    return accumulator;
}

int part2(const std::vector<std::vector<unsigned short>>& reports){
    unsigned int accumulator = 0;
    for (const std::vector<unsigned short>& report : reports){
        std::tuple<bool, unsigned int> result = is_report_safe(report);
        if (std::get<bool>(result)){
            accumulator++;
        }
        else {
            std::vector<unsigned short> dampened_report1 = report;
            dampened_report1.erase(dampened_report1.begin() + std::get<unsigned int>(result));
            std::vector<unsigned short> dampened_report2 = report;
            dampened_report2.erase(dampened_report2.begin() + std::get<unsigned int>(result) + 1);
            
            if (std::get<bool>(is_report_safe(dampened_report1)) or std::get<bool>(is_report_safe(dampened_report2))){
                accumulator++;
            }
        }
    }
    return accumulator;
}

int main(int argc, char *argv[]){
    std::vector<std::vector<unsigned short>> reports;
    
    std::ifstream input;
    input.open("day2.txt");
    std::string line;
    while (std::getline(input, line)){
        std::vector<unsigned short> report;
        std::stringstream ss(line);
        std::string item;
        while(std::getline(ss, item, ' ')){
            report.push_back(std::stoi(item));
        }
        reports.push_back(report);
    }
    input.close();

    std::cout << part1(reports) << std::endl;
    std::cout << part2(reports) << std::endl;
    
    return 0;
}